api/src/main/java/com/github/retrooper/packetevents/wrapper/play/server/WrapperPlayServerMapData.java

```diff
@@ -0,0 +1,177 @@
+                                    byte[] buffer,
+                                    int x, int z,
+                                    int column, int rows) {
+        super(PacketType.Play.Server.MAP_DATA);
+        this.data = data;
+        this.scale = scale;
+        this.locked = locked;
+        this.icons = icons;
+        this.buffer = buffer;
+        this.column = column;
+        this.rows = rows;
+        this.x = x;
+        this.z = z;
+    }
+
+    @Override
+    public void read() {
+        this.data = this.readVarInt();
+        this.scale = this.readByte();
+
+        boolean readIcons = true;
+        if (this.clientVersion.isNewerThanOrEquals(ClientVersion.V_1_9)) {
+            if (this.clientVersion.isOlderThan(ClientVersion.V_1_17)) {
+                readIcons = this.readBoolean();
+                this.locked = this.readBoolean();
+            } else {
+                this.locked = this.readBoolean();
+                readIcons = this.readBoolean();
+            }
+        }
+
+        if (readIcons || this.clientVersion.isOlderThan(ClientVersion.V_1_17)) {
+            int iconCount = this.readVarInt();
+            this.icons = new ArrayList<>(iconCount);
+            for (int i = 0; i < iconCount; i++) {
+                this.icons.add(new Icon(this));
+            }
+        }
+
+        this.column = this.readUnsignedByte();
+        if (this.column > 0) {
+            this.rows = this.readUnsignedByte();
+            this.x = this.readUnsignedByte();
+            this.z = this.readUnsignedByte();
+            this.buffer = this.readByteArray();
+        }
+    }
+
+    @Override
+    public void write() {
+        this.writeVarInt(this.data);
+        this.writeByte(this.scale);
+
+        boolean writeIcons = true;
+        if (this.clientVersion.isNewerThanOrEquals(ClientVersion.V_1_9)) {
+            writeIcons = this.icons != null;
+            if (this.clientVersion.isOlderThan(ClientVersion.V_1_17)) {
+                this.writeBoolean(writeIcons);
+                this.writeBoolean(this.locked);
+            } else {
+                this.writeBoolean(this.locked);
+                this.writeBoolean(writeIcons);
+            }
+        }
+
+        if (writeIcons || this.clientVersion.isOlderThan(ClientVersion.V_1_17)) {
+            this.writeVarInt(this.icons.size());
+            this.icons.forEach(icon -> {
+                this.writeByte((icon.getType() & 15) << 4 | icon.getRotation() & 15);
+                this.writeByte(icon.getX());
+                this.writeByte(icon.getY());
+            });
+        }
+
+        this.writeByte(this.column);
+        if (this.column > 0) {
+            this.writeByte(this.rows);
+            this.writeByte(this.x);
+            this.writeByte(this.z);
+            this.writeByteArray(this.buffer);
+        }
+    }
+
+    @Override
+    public void copy(WrapperPlayServerMapData wrapper) {
```

Comment:
No need to call `super.copy` or clone the buffer. Only copy local values

Reply:
removed

Reply:
this.buffer = wrapper.buffer.clone(); - this too ^^

Reply:
done!

Reply:
I actually got confused with the buffer.. my bad. This one has to be copied as you provide it as a parameter. You should probably give it a different name because we are using bytebuf/buffer getters in the PacketWrapper class.